### PR TITLE
Implemented a 'choose' permission for Images and Documents

### PR DESCRIPTION
**Pull request by coredumperror**
This PR adds the 'choose' permission discussed in #4488. It limits Image and Document choosers to showing only those Collections (and their contents) for which the current user has the Choose permission.

It grants Choose permissions on the Root Collection to every existing Group that has the "Can access Wagtail admin" perm, via data migration. So there will be no change from previous behavior until the user goes in and alters the Choose perms for their site's Groups.

This PR has no effect on anything except choosers. The Image and Document index pages are still controlled solely by the 'add' and 'change' permissions, as those are only accessible by users who have those permissions in the first place. This means that it's now possible to configure a Group to allow its members to choose images from only specific Collections when editing pages, *without* them being able to edit said images or add new ones to said Collections.

However, this does add one wrinkle that I'd like to fix, but I'm not entirely sure how. I'd like for the `AdminImageChooser` and `AdminDocumentChooser` widgets to hide the "Edit this image/document" button if the user only has Choose permission. But since low level stuff like the widget constructor code doesn't have access to the Django request, I'm not sure how to actually check if the current user has said permission, because I have no access to the current user. There's no obvious way I can see to pass the request (or just the user) from the view down into the widget constructor (or the `render_html()` method).

Anyway, I've written the code that will be able to implement edit button hiding, *if* I can find a way to get the User object. But since I don't know how, it's commented out.

I will add docs and tests for the new choose permission functionality in the next commit. One existing test fails due to my changes, which I will fix ASAP.

There are also several linter errors, but only in files I didn't touch, so I'm going to leave them alone.

### CODE REVIEW

**Comment by coredumperror** [comment]
OK, new commit fixes the lint errors by merging with the newest master, and fixes the error with the existing test by having that test code add the necessary Choose permission.

**Comment by coredumperror** [comment]
By the way, I discovered a bug in how the Group Collection permissions formsets are rendered.

I discovered the issue because of how I originally wrote the changes to `documents/forms.py` and `images/forms.py`. `('choose_image', _("Choose"), _("Select images in Choosers")),` was originally at the top of the permissions list, but that caused a really baffling bug in the form. It displayed the checkboxes as "Choose   Add   Edit", but when you actually checked them and saved the form, it would save the permissions as if the order was "Add   Edit   Choose".

Thus, checking "Choose" would actually give the "Add" permission. Figuring out why the heck this was happening was a frustrating process, but I eventually hypothesized that the issue might be due to alphabetical storing, so I put "choose_image" *after* "Edit" (which is actually "change_image"), and that fixed it.

My guess is that this has always been a bug in `collection_member_permission_formset_factory()`, but it's never manifested because the only two permissions that existed previously were coincidentally placed in alphabetical order. I believe the perpetrator is the first code in that function: 
```    
    permission_queryset = Permission.objects.filter(
        content_type__app_label=model._meta.app_label,
        codename__in=[codename for codename, short_label, long_label in permission_types]
    ).select_related('content_type')
```

This ends up making the form code ignore the order from `permission_types` and instead use the default ordering of the `Permission` model, which orders permissions for the same model in "codename" order. However, the template that renders the checkbox's labels gets them directly from `permission_types`. Thus, the order is preserved for the table headers, but not the checkboxes.

Since this PR is why this bug was discovered, should I fix it in this PR? Or should it be fixed in a separate one? For now, the PR can work despite the bug, by just putting "Choose" at the end of the list of checkboxes. Though it does feel more natural to put it at the start.

**Comment by ababic** [comment]
@coredumperror Thanks very much for getting going with this. This is valuable start, but I still think there are some conversations to be had about how we this should work.

1.  Should a user need the the 'choose' permission explicitly in order to choose from a Collection?

If a user has the `create` or `change` permission for a collection (either directly, or via permissions set on an ancestor), couldn't we assume the `choose` permission? This would make the 'choose' permission a bit of a special case (not unlike Django's recently added 'view' permission), in that it is a lesser permission than `create` or `change`, that is only really ever meaningfully granted to users who aren't trusted enough to wield 'manage' the media. I personally feel this is okay, but would am keen to hear other points of view.

2. Should all of the functionality behind the 'Images' and 'Documents' menu items be reserved for those with the `create` or `change` permission for a collection, or would there be value in a 'read-only' version of those views?

This isn't necessarily something that would need to be done in the same PR, but in Layman's terms, by granting the 'choose' permission, aren't we essentially saying that users in that group can "See that the collection exists" and "Use media from the collection". When we think about that user's experience within the CMS, are there things that could be useful for them to 'see' in regards to that media, that might not be surfaced in the 'choose' UI? And is `choose` the correct verb for this permission after all (perhaps `view` or `browse` makes more sense)?

File: wagtail/documents/migrations/0011_add_choose_permissions.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    Permission = apps.get_model('auth.Permission')
+    document_content_type = ContentType.objects.get(
+        model='document',
+        app_label='wagtaildocs',
+    )
+    # This cascades to Group
+    Permission.objects.filter(
+        content_type=document_content_type,
+        codename='choose_document'
+    ).delete()
```
Comment (ababic) [inline wagtail/documents/migrations/0011_add_choose_permissions.py:43]: It's fine to create the permission to use in the data migration (as usually, it wouldn't be created until Django's 'post_migrate' signal is emitted), but I don't think we want to DELETE it here. We're better off letting Django take care of that.

**Comment by ababic** [comment]
Should officially define the new permission on the `Meta` class of `Image` and `Document`? This way, if for any reason the `Permission` object is deleted, Django should automatically detect that and recreate it. Or, perhaps we can just reuse the `view_image` and `view_document` permissions that Django will have created automatically for these models?

File: wagtail/documents/migrations/0011_add_choose_permissions.py (hunk 1 above)
Comment (coredumperror) [reply wagtail/documents/migrations/0011_add_choose_permissions.py:43]: I just copied what the images and documents apps do with the original `add_document`, `change_image`, etc. permissions in their respective backwards migrations. See `wagtail/images/migrations/0001_squashed_0021.py` and `wagtail/documents/migrations/0002_initial_data.py`. I wrote this PR's migration based on what those two files do.

As for "letting Django take care of that", what do you mean? In my experience, Django is notorious for *not* deleting permissions for models that have been removed from the database.

**Comment by coredumperror** [comment]
> Should we officially define the new permission on the Meta class of `Image` and `Document`?  

The Image and Document models don't currently define the Add and Edit permissions in their Metas. Should those be defined there, too? That's probably the "right" way to do it, but I'm not sure how it'd interact with established Wagtail projects. And defining permissions in multiple different ways sounds like a disaster waiting to happen.

> Should a user need the 'choose' permission explicitly in order to choose from a Collection?

Having the Add or Edit permissions act as a proxy for the Choose permission just sounds confusing to me, now that a separate Choose permission actually exists. An admin with permission to edit Groups might look at a Group's perms and say "Wait a minute, Suzy told me she can choose images from this Collection, but her group doesn't have the Choose perm!"

Sure, having Edit and _not_ Choose would be weird a weird state, but this PR already gives everyone the Choose permission on the Root Collection (to retain existing behavior), so why would anyone remove Choose but leave Add or Edit?

> Should all of the functionality behind the 'Images' and 'Documents' menu items be reserved for those with the add or change permission for a collection, or would there be value in a 'read-only' version of those views?

I considered this, but treated it as outside the scope of this PR. I figure that the entire purpose of the Index views for Images and Documents are for adding new ones and editing existing ones. We'd need essentially a whole new "read only index view" for people who have Choose but neither Edit nor Add, and we'd want to make it really clear that the read-only one a completely different from the add/edit one, or users are likely to get quite confused. 

Plus, what about the case where Suzy has add/edit permission Collection 2, but *only* Choose on Collection 3 (which is a sibling of Collection 2)? The Index view would need to show Images in Collection 3 as "read only" but images from Collection 2 as editable, even on the "view all images" version of the index. That'd get real confusing, real fast.

This is one of the reasons that I figure "Choose" is the right name for this permission, because it's *specifically* for Choosers. It is not affiliated with the Index pages at all, so there's no way for confusion to occur on the Index pages for people with disparate permissions on separate Collections. 

This also ties back into the "Should a user need the 'choose' permission explicitly in order to choose from a Collection?" question. This PR takes the implied "Choose" permission out of the Edit and Add perms and makes it a completely separate thing.

**Comment by coredumperror** [comment]
It's been a while since the last comment around here. I'm really hoping we can get this PR integrated into Wagtail 2.12, so if there's anything I need to do to help that happen, please let me know. :)

**Review by gasman — APPROVED** [review]
> > Should we officially define the new permission on the Meta class of `Image` and `Document`?
> 
> The Image and Document models don't currently define the Add and Edit permissions in their Metas. Should those be defined there, too? That's probably the "right" way to do it, but I'm not sure how it'd interact with established Wagtail projects. And defining permissions in multiple different ways sounds like a disaster waiting to happen.

Looking at https://docs.djangoproject.com/en/3.1/topics/auth/default/#default-permissions and https://docs.djangoproject.com/en/3.1/topics/auth/customizing/#custom-permissions , Django makes a distinction between 'default permissions' (add, edit, delete) and custom permissions, so based on that, defining the choose permission (but not the others) in Meta would fit with how Django expects things to work. It's probably not strictly necessary given the `get_or_create` in the migration, but it can't hurt, in case there is some unforeseen mechanism that garbage-collects them (see [remove_stale_contenttypes](https://github.com/wagtail/wagtail/issues/4444)...)

> > Should a user need the 'choose' permission explicitly in order to choose from a Collection?
> 
> Having the Add or Edit permissions act as a proxy for the Choose permission just sounds confusing to me, now that a separate Choose permission actually exists. An admin with permission to edit Groups might look at a Group's perms and say "Wait a minute, Suzy told me she can choose images from this Collection, but her group doesn't have the Choose perm!"
> 
> Sure, having Edit and _not_ Choose would be weird a weird state, but this PR already gives everyone the Choose permission on the Root Collection (to retain existing behavior), so why would anyone remove Choose but leave Add or Edit?

I agree - to quote the Zen of Python, "explicit is better than implicit". Even if there's no obvious reason why an administrator might give a user add/edit but not choose permission, the best thing Wagtail can do in that situation is to behave as instructed rather than try to second-guess what they meant.

I'm happy with the overall approach here, and with the Meta permission lines added and the small tweaks below addressed, I'm happy to approve this. Would be good to find a fix for the checkbox ordering bug - I agree that it feels more natural for 'choose' to be the first option - but that shouldn't be a blocker to getting this merged.

**Comment by gasman** [comment]
Merged in d18ffb0233dfb01c5fc876fa7e9b755aa9a256f0, with the proposed fixes applied and some additional documentation. Thanks @coredumperror!

**Comment by coredumperror** [comment]
Woooo!! 🥳
